distributed/scheduler.py

```diff
@@ -5894,31 +5966,11 @@ async def _rebalance_move_data(
             },
         )
 
-        if any(r["status"] != "OK" for r in result):
-            return {
-                "status": "missing-data",
-                "keys": list(
-                    concat(
-                        r["keys"].keys()
-                        for r in result
-                        if r["status"] == "missing-data"
-                    )
-                ),
-            }
-
-        for snd_ws, rec_ws, ts in msgs:
-            assert ts._state == "memory"
-            ts._who_has.add(rec_ws)
-            rec_ws._has_what[ts] = None
-            rec_ws.nbytes += ts.get_nbytes()
-            self.log.append(
-                ("rebalance", ts._key, time(), snd_ws.address, rec_ws.address)
-            )
-
-        await asyncio.gather(
-            *(self._delete_worker_data(r, v) for r, v in to_senders.items())
-        )
-        return {"status": "OK"}
+        missing_keys = {k for r in failed_keys_by_recipient.values() for k in r}
+        if missing_keys:
+            return {"status": "missing-data", "keys": list(missing_keys)}
```

Comment:
This is currently forwarded to the user/caller of replicate/rebalance, correct? We should probably document what exactly missing-data in this context means. IIUC this does not necessarily mean that the task is lost but merely that it was not where we expected it to be, for whatever reason.

Reply:
I've changed Client.rebalance to raise KeyError if it receives missing-data AND the client explicitly listed futures to be rebalanced

Reply:
Correction: the logic described above is actually in Scheduler.rebalance